python/cutlass_library/generator.py

```diff
@@ -2836,6 +2836,151 @@ def GenerateSM80_TensorOp_16832_TN(manifest, cuda_version):
       else:
         op.C.alignment = 8
 
+#
+def GenerateSM80_TensorOp_16832_TN_mixed_input_upcast_a(manifest, cuda_version):
+
+  if not CudaToolkitVersionSatisfies(cuda_version, 11, 0):
+    return
+
+  layouts = [
+    (LayoutType.RowMajor, LayoutType.ColumnMajor, LayoutType.ColumnMajor),
+  ]
+
+  # Upcast on Operand A
+  math_instructions = [
+    MathInstruction(                                  \
+      [16, 8, 32],                                    \
+      DataType.s4, DataType.s8, DataType.s32,         \
+      OpcodeClass.TensorOp,                           \
+      MathOperation.multiply_add_mixed_input_upcast),
+  ]
+
+  min_cc = 80
+  max_cc = 1024
+
+  # For mixed-input alignment constraints are a list of lists, where the 
+  # inner list contains the alignment constraints for operands/matrices 
+  # [[alignA, alignB, alignC],..]
+  alignment_constraints = [[32, 16, 4],]
+
+  for math_inst in math_instructions:
+    tile_descriptions = [
+      TileDescription([256, 128,  64],  3, [4, 2, 1], math_inst, min_cc, max_cc),
+      TileDescription([128, 256,  64],  3, [2, 4, 1], math_inst, min_cc, max_cc),
+      TileDescription([256,  64,  64],  4, [4, 1, 1], math_inst, min_cc, max_cc),
+      TileDescription([ 64, 256,  64],  4, [1, 4, 1], math_inst, min_cc, max_cc),
+      TileDescription([ 32, 256,  64],  4, [1, 4, 1], math_inst, min_cc, max_cc),
+      TileDescription([128, 128,  64],  5, [2, 2, 1], math_inst, min_cc, max_cc),
+      TileDescription([ 64, 128,  64],  6, [2, 2, 1], math_inst, min_cc, max_cc),
+      TileDescription([256, 128, 128],  3, [4, 2, 1], math_inst, min_cc, max_cc),
+      TileDescription([128, 256, 128],  3, [2, 4, 1], math_inst, min_cc, max_cc),
+      TileDescription([256,  64, 128],  4, [4, 1, 1], math_inst, min_cc, max_cc),
+      TileDescription([ 64, 256, 128],  4, [1, 4, 1], math_inst, min_cc, max_cc),
+      TileDescription([256,  32, 128],  4, [4, 1, 1], math_inst, min_cc, max_cc),
+      TileDescription([ 32, 256, 128],  4, [1, 4, 1], math_inst, min_cc, max_cc),
+      TileDescription([128, 128, 128],  4, [2, 2, 1], math_inst, min_cc, max_cc),
+      TileDescription([ 64, 128, 128],  3, [2, 2, 1], math_inst, min_cc, max_cc),
+      TileDescription([128,  32, 128],  4, [4, 1, 1], math_inst, min_cc, max_cc),
```

Comment:
How did you come up with this `TileDescription` list for `S8` x `S4`? I guess you carried these from `S8` x `S8`. Please make sure all of these pass verification. You can follow steps similar to [here](https://github.com/NVIDIA/cutlass/pull/1350#issuecomment-2223601983) to instantiate all the tile shapes listed here by using `-DCUTLASS_LIBRARY_KERNELS="s8_s4,s4_s8"`. By default the build process only instantiate 128x128 tile shape.

Reply:
Can you please run the verification and profiling on `--m=3456 --n=4096 --k=2048` on an A100? 

Please compile using `-DCUTLASS_LIBRARY_KERNELS="s8_s4,s4_s8,s4,s8"` to also have s4 x s4 and s8 x s8 kernels in the runs.

Reply:
The tiles selection is desribed in a [comment](https://github.com/NVIDIA/cutlass/pull/1413#issuecomment-2015366925) above; also, as mentioned in this comment, I did the verification.  I will repeat the verification procedure, together with profiling, and report the outcome here.

Reply:
There was a build issue after rebasing on the latest main: basically, `OpMultiplyAddSaturate` for `MmaTensorOpPolicy` in the specialization of `struct DefaultMmaTensorOp` (in `include/cutlass/gemm/warp/default_mma_tensor_op_sm80.h`) seem to be obligatory now, as the build fails if `OpMultiplyAdd` used.  The branch is updated accordingly.

I've configured the build using `-DCUTLASS_NVCC_ARCHS=80 -DCUTLASS_LIBRARY_KERNELS="s8_s4,s4_s8,s4,s8"` CMake options, and then verified that `cutlass_test_unit_gemm_device_mixed_input_tensorop_sm80` unit test passes.  Then, I did profiler runs as follows:
```
./build/tools/profiler/cutlass_profiler --operation=gemm --m=3456 --n=4096 --k=2048 --A=s8:row --B=s8:column >& s8_s8.txt
./build/tools/profiler/cutlass_profiler --operation=gemm --m=3456 --n=4096 --k=2048 --A=s4:row --B=s8:column >& s4_s8.txt
./build/tools/profiler/cutlass_profiler --operation=gemm --m=3456 --n=4096 --k=2048 --A=s8:row --B=s4:column >& s8_s4.txt
```

The corresponding profiler outputs are here: 
[s8_s8.txt](https://github.com/user-attachments/files/16630281/s8_s8.txt)
[s4_s8.txt](https://github.com/user-attachments/files/16630289/s4_s8.txt)
[s8_s4.txt](https://github.com/user-attachments/files/16630290/s8_s4.txt)

The disposition values for mixed data types cases with `s8` accumulator are still `incorrect`.  Also, the timings are somewhat slower than for corresponding `s8xs8` cases (with the same configurations: tile sizes etc.).

Reply:
Thank you for running and sharing these results. 

Accumulator is for all of these runs should be `S32` as shown at the bottom of the output in csv format with `accum` type = `S32`. The `Incorrect` disposition with mixed-input is happening for only `S8` output, i.e., when the accumulators are `S32` but the output is downcast-ed to `S8`. 

We do *not* see incorrect results for `S8`x`S8` with `S32` accumulators and `S8` output, can you pick one row of incorrect run from 

```
(elementD/elementC type) <= (elementA type) x(elementB type) + (accum type)
S8  <= S8  x S4  + S32 
```

and compare the same kernel configuration against 

```
S8 <= S8 x S8 + S32
```
to find where is the difference?


I believe it is to do with initialization of the operands during profiling or inside the kernel epilogue S32-to-S8. quantization. 

Also, you can just upload the csv that can be produced by adding `--output=filename.csv` to the profiler runs

Reply:
The CSV files are attached below, I'm examining `incorrect` cases.

[s8_s8.gemm.csv](https://github.com/user-attachments/files/16635796/s8_s8.gemm.csv)
[s4_s8.gemm.csv](https://github.com/user-attachments/files/16635801/s4_s8.gemm.csv)
[s8_s4.gemm.csv](https://github.com/user-attachments/files/16635803/s8_s4.gemm.csv)

Reply:
Here is what I found so far regarding `incorrect` cases:

First, I made following change in the code generating inputs, in order to generate the same inputs for profiler for `S4xS8` and `S8xS8` cases:
```
diff --git a/tools/profiler/src/device_context.cu b/tools/profiler/src/device_context.cu
index 2cbfa5d2..7b488fe8 100644
--- a/tools/profiler/src/device_context.cu
+++ b/tools/profiler/src/device_context.cu
@@ -105,7 +105,7 @@ DeviceAllocation *DeviceContext::allocate_tensor(
           data_distribution.set_uniform(-1, 1, 0);
           break;
         case library::NumericTypeID::kS4:
-          data_distribution.set_uniform(-2, 2, 0);
+          data_distribution.set_uniform(-3, 3, 0);
           break;
         case library::NumericTypeID::kU2:
           data_distribution.set_uniform(0, 2, 0);
```

I used following profiler runs to make comparision between `S4xS8` and `S8xS8` cases (BTW, I found that smaller input shapes selection that would still allow for reproducing the problem would be `--m=32 --n=64 --k=512`):
```
cutlass_profiler --operation=gemm --gemm_kind=universal --m=3456 --n=4096 --k=2048 --A=s8:row --B=s8:column --C=s8:column --D=s8:column --accum=s32 --cta_m=256 --cta_n=128 --cta_k=64 --stages=3 --save-workspace=always
cutlass_profiler --operation=gemm --gemm_kind=universal --m=3456 --n=4096 --k=2048 --A=s4:row --B=s8:column --C=s8:column --D=s8:column --accum=s32 --cta_m=256 --cta_n=128 --cta_k=64 --stages=3 --save-workspace=always
```

By comparing saved `.mat` files, I verified that input matrices `A`, `B` and `C` are the same, but also that output matrices `D` are the same.  What differs are actually `Reference` matrices, which means that reference results calculated for `S4xS8` case are wrong.  If I understood it correctly, cuBLAS is used for reference calculations, so I'll check what's going on there...

Reply:
I am not sure if cuBLAS is called for this for reference check. The output should show which references are called. You can use `--verification-providers=cublas,host,device` to run them all. Is there a host reference for this you must check in here [/tools/library/src/reference](/tools/library/src/reference ) ?

Reply:
Indeed - `device` provider is actually used for reference check here.  I posted an update with the fix in reference calculations, so for most of cases with `S8` output, `cutlass_profiler` reports success now.  However, there are still couple of cases where `incorrect` is reported, I'm looking into this...

Reply:
Pushed another update - the problem with remaining `incorrect` cases was that I haven't copied `C` operand alignment update from `S8xS8` case, in the generator code.  Everything is reported as `passed` now by profiler, the output files are attached below.  I believe this one should be ready for merging now.

[s8_s8.gemm.csv](https://github.com/user-attachments/files/16681357/s8_s8.gemm.csv)
[s4_s8.gemm.csv](https://github.com/user-attachments/files/16681359/s4_s8.gemm.csv)
[s8_s4.gemm.csv](https://github.com/user-attachments/files/16681361/s8_s4.gemm.csv)

Reply:
@alexsamardzic thank you for digging it through. LGTM!

@hwu36 , @thakkarV , @IonThruster , can you please help it merge it?